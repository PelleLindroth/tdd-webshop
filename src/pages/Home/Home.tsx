import { useEffect, useState } from 'react'
import { IProduct } from '../../mocks/products'
import Card from '../../components/Card/Card'

const Home = (props: { products: IProduct[] }) => {
  const { products } = props
  const [searchPhrase, setSearchPhrase] = useState('')
  const [filteredProducts, setFilteredProducts] = useState<IProduct[]>([])

  useEffect(() => {
    setFilteredProducts(
      products.filter((product) =>
        product.name.toLowerCase().includes(searchPhrase)
      )
    )
  }, [products, searchPhrase])

  const handleSearch = (e: React.FormEvent<HTMLInputElement>) => {
    const phrase = e.currentTarget.value.toLowerCase()
    setSearchPhrase(phrase)
  }

  return (
    <main>
      <h1>HOME</h1>
      <div>
        <label htmlFor="search">Search product:</label>
        <input
          onChange={handleSearch}
          value={searchPhrase}
          id="search"
          type="search"
        />
      </div>
      {products.length === 0 && <h3>Something went wrong</h3>}
      {filteredProducts.length > 0 ? (
        <ul>
          {filteredProducts.map((product) => (
            <Card product={product} key={product.id} />
          ))}
        </ul>
      ) : (
        <h3>No matching products found</h3>
      )}
    </main>
  )
}

export default Home
