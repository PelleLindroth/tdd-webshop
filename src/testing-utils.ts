import { render } from '@testing-library/react'
import { ReactElement } from 'react'
import { MemoryRouter } from 'react-router-dom'

const renderWithRouter = (ui: ReactElement, options?: any) => render(ui, { wrapper: MemoryRouter, ...options })

export { renderWithRouter }
