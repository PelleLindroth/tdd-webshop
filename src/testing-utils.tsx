import { render } from '@testing-library/react'
import { ReactElement } from 'react'
import { MemoryRouter } from 'react-router-dom'
import { Routes, Route } from 'react-router'

const renderWithRouter = (ui: ReactElement, options?: any) =>
  render(ui, { wrapper: MemoryRouter, ...options })

const withPath = (entry: string, element: ReactElement, path: string) => (
  <MemoryRouter initialEntries={[entry]}>
    <Routes>
      <Route path={path} element={element} />
    </Routes>
  </MemoryRouter>
)
export { renderWithRouter, withPath }
