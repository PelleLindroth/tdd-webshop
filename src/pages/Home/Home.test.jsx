import { shallow } from 'enzyme'
import { screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import Home from './Home'
import { renderWithRouter } from '../../testing-utils'

describe('Home', () => {
  it('renders Home component correctly', () => {
    const wrapper = shallow(<Home />)

    expect(wrapper).toMatchSnapshot()
  })
  it('renders a list of 6 products initially', () => {
    renderWithRouter(<Home />)
    const products = screen.getAllByRole('listitem')
    expect(products).toHaveLength(6)
  })
  it('renders an empty input field', () => {
    renderWithRouter(<Home />)
    const input = screen.getByRole('searchbox')
    expect(input).toHaveTextContent('')
  })
  it('shows correct filtered products when searching by product name', () => {
    renderWithRouter(<Home />)
    const input = screen.getByRole('searchbox')

    const filteredProducts = screen.getAllByRole('listitem')

    expect(filteredProducts).toHaveLength(6)

    userEvent.type(input, 'book')

    const newlyFilteredProducts = screen.getAllByRole('listitem')

    expect(newlyFilteredProducts).toHaveLength(1)
  })
  it('returns car when CAR is searched (case insentitive search)', () => {
    renderWithRouter(<Home />)
    const input = screen.getByRole('searchbox')
    userEvent.type(input, 'CAR')
    const newlyFilteredProducts = screen.queryAllByRole('listitem')

    expect(newlyFilteredProducts).toHaveLength(1)
  })
  it('shows no filtered products if search doesnt match', () => {
    renderWithRouter(<Home />)
    const input = screen.getByRole('searchbox')

    const filteredProducts = screen.getAllByRole('listitem')

    expect(filteredProducts).toHaveLength(6)

    userEvent.type(input, 'kljhter4k')

    const disappearedProducts = screen.queryAllByRole('listitem')

    expect(disappearedProducts).toHaveLength(0)
  })
  it('shows message if no products match search', () => {
    renderWithRouter(<Home />)
    const input = screen.getByRole('searchbox')

    userEvent.type(input, 'kljhter4k')

    const message = screen.getByText(/no matching products/i)

    expect(message).toBeInTheDocument()
  })
})
