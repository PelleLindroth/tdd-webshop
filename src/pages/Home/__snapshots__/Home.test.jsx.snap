// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders Home component correctly 1`] = `
<main>
  <h1>
    HOME
  </h1>
  <div>
    <label
      htmlFor="search"
    >
      Search product:
    </label>
    <input
      id="search"
      onChange={[Function]}
      type="search"
      value=""
    />
  </div>
  <h3>
    No matching products found
  </h3>
</main>
`;
